src/main/java/org/jenkinsci/backend/ircbot/IrcListener.java

```diff
@@ -466,79 +484,56 @@ private void replyBugStatus(String channel, String ticket) {
      *
      * IOW, does he have a voice of a channel operator?
      */
-    private boolean isSenderAuthorized(String channel, String sender) {
+    private boolean isSenderAuthorized(Channel channel, User sender) {
         return isSenderAuthorized(channel, sender, true);
     }
     
-    private boolean isSenderAuthorized(String channel, String sender, boolean acceptVoice) {
-        for (User u : getUsers(channel)) {
-            System.out.println(u.getPrefix()+u.getNick());
-            if (u.getNick().equals(sender)) {
-                String p = u.getPrefix();
-                if (p.contains("@") || (acceptVoice && p.contains("+")) 
-                        || (IrcBotConfig.TEST_SUPERUSER != null && IrcBotConfig.TEST_SUPERUSER.equals(sender) ))
-                    return true;
-            }
-        }
-        return false;
-    }
-
-    @Override
-    protected void onDisconnect() {
-        while (!isConnected()) {
-            try {
-                reconnect();
-                for (String channel : IrcBotConfig.getChannels()) {
-                    joinChannel(channel);
-                }
-            } catch (Exception e) {
-                e.printStackTrace();
-                try {
-                    Thread.sleep(15000);
-                } catch (InterruptedException _) {
-                    return; // abort
-                }
-            }
-        }
+    private boolean isSenderAuthorized(Channel channel, User sender, boolean acceptVoice) {
+        return sender.getUserLevels(channel).stream().anyMatch(e -> e == UserLevel.OP || (acceptVoice && e == UserLevel.VOICE)
```

Comment:
Not the best part of Java 8 IMHO. LINQ--

Reply:
You hurt my feelings as a .NET Dev ;-)